source/develop/release-management/features/network/allowExplicitDnsConfiguration.html.md

```diff
@@ -0,0 +1,132 @@
+authors: mmucha
+wiki_category: Feature
+wiki_title: Features/AllowExplicitDnsConfiguration
+wiki_revision_count: 12
+wiki_last_updated: 2015-10-14
+feature_name: Allow Explicit Dns Configuration
+feature_modules: Networking
+feature_status: Design
+rfe: https://bugzilla.redhat.com/show_bug.cgi?id=1160667
+
+---
+
+# Allow Explicit Dns Configuration
+
+### Owner
+
+*   Name: [ Martin Mucha](User:mmucha)
+*   Email: mmucha@redhat.com
+
+## Summary
+When new host is added to the system, management network will be created for it. Dns configuration for such network will be obtained from resolv.conf file. However admin should be able to specify overriding configuration in WebAdmin. He can do that at several places:
+
+* During creation new host
+* By updating management network of given host (in backend terms — via its NetworkAttachment)
+* By updating management network of given DC (or later of given Cluster) 
+
+## Detailed Description
+
+So DNS configuration can be stored in multiple places. If DNS configuration is not specified in either of them, no DNS configuration data will be passed to VDSM during SetupNetworks command and default values from resolv.conf will be used. If either of them is used, it will be passed to VDSM overriding configuration from resolv.conf. For example we can set DNS configuration in Network related to certain DC (or later to certain Cluster) and this configuration will override configuration in resolv.conf. If we now setup configuration in NetworkAttachment, attaching ManagementNetwork of given DataCenter (or later of given Cluster) to specific Host, this configuration is more specific than one in Network of DC (or later Cluster) scope and will be used instead.  
+
+### DB
+Database needs to be updated so it can accommodate dns configuration. Two places will be altered:
+
+* network_attachments table
+* network table
+
+both this tables must accommodate multiple DNS records. Naive solution would be to store them comma separated into single column. Better solution would be creating separate table for that. That's subject of further discussion.
+
+### REST
+
+As mentioned, you can specify DNS Configuration at three places. Corresponding REST areas will be altered. But first we need to add new element 'dns_configuration' as: 
+
+```
+<xs:element name="dns_configuration" type="DnsConfiguration"/>
+
+<xs:complexType name="DnsConfiguration">
+  <xs:complexContent>
+    <xs:extension base="BaseResource">
+      <xs:sequence>
+        <xs:element maxOccurs="1" minOccurs="0" name="ips" type="Ips"/>
+      </xs:sequence>
+    </xs:extension>
+  </xs:complexContent>
+</xs:complexType>
+```
+
+#### Creating new Host
+```
+POST /ovirt-engine/api/hosts HTTP/1.1
+Content-Type: application/xml
+
+<host>
+    <name>$NAME</name>
+    <address>$ADDRESS</address>
+    <port>$PORT</port>>
+    <root_password>$PASSWORD</root_password>
+    <cluster>
+        <name>$CLUSTERNAME</name>
+    </cluster>
+    <dns_configuration>
```

Comment:
We need a mechanism to preserve backwards compatibility of the API. Currently when a host is added (with `POST /hosts`) its DNS resolver configuration isn't touched, if I understand correctly. Users may rely on that: an administrator may have manually and carefully configured the `resolv.conf`file, before adding the host, and these changes would be lost without warning if the management network has a custom DNS resolver configuration. That is the objective of the `preseve_dns_resolver_configuration`, keep backwards compatibility and still be able to use the new behavior.

We can't include the DNS resolver configuration in the request to add a host unless it becomes an attribute of the `Host` type. When you do a `POST` to the hosts collection the body of the request is the XML (or JSON) representation of the API `Host` type. So if we want to be able to send this:

```xml
<host>
  ...
  <dns_resolver_configuration>
    ...
  </dns_resolver_configuration>
</host>
```

Then we have no alternative but changing the definition `Host` type in the specification of the API like this:

```java
@Type
public interface Host {
  ...
  /**
   * Blah, blah, blah ...
   */
  DnsResolverConfiguration dnsResolverConfiguration();
  ...
}
```

If we do that then the generated documentation and the SDKs will list `dns_resolver_configuration` as just another attribute of the `Host` type. Same will happen for all the SDKs. Users will then expect to be able to use it when adding, retrieving and updating hosts. But that won't be true, it will only be available when adding hosts. That is confusing, at least. We could alleviate the problem with documentation:

```java
/**
 * This attribute only makes sense when adding a host. It won't be populated when
 * retrieving a host, and it will be ignored when updating.
 */
DnsResolverConfiguration dnsResolverConfiguration();
```

Will users ever read that? Most won't.

So the best solution is to either make this a real attribute, that works for adding, retrieving and updating, or else not have it at all.

It is OK to have it as an attribute of the `Network` type, assuming that it works for adding, retrieving and updating networks.

It is OK to have it as an attribute of the `NetworkAttachment` type, again assuming that it works for adding, retrieving and updating network attachments.

For `setupnetworks` it is OK to have it as attributes of the network attachments if the meaning is that the provided value will be part of the created network attachment. If the provided value will only be used to update the host and won't be stored as part of the created network attachment, then it is better to have it as parameter of the operation.